### ISSUE TITLE
[2.7][Serializer] Groups skipping array values instead of throwing a circular reference error

### ISSUE THREAD

**Issue by soyuka** [symfony/symfony#14495]
Hi,

Since Symfony Live 2015 and the @dunglas slides about [ApiPlatform](http://dunglas.fr/2015/04/mes-slides-du-symfony-live-2015-a-la-decouverte-de-api-platform/), I've read a lot on the future 2.7 serializer and wanted to give it a shot (MIT licence ftw).

I've replaced JMSSerializer by the Symfony serializer (v2.7). FIY I'm using FOSRest, FOSUser. Everything works as expected except for many-to-many relationships that leads to the circular reference error : 
```
A circular reference has been detected (configured limit: 1).
```

It's easy to reproduce by using groups in a FOSUser class for example ([gist of my use case](https://gist.github.com/soyuka/5cd32901041f2fbf452c)).

I was wondering, is this an expected behavior or is it something that still needs a fix? Should I implement a repository that returns the result of the many-to-many relation or should the Serializer do the job by detecting the correct entity metadata?

Thanks!

**Comment by dunglas** [comment]
Hi @soyuka,

The serializer don't care about Doctrine and internal join table. Can you provide the `UserGroup` entity as well?

If I understand well there is a circular reference in your exemple because of something like a `users` property in the `UserGroup` class: User -> UserGroup -> User -> UserGroup...
The easiest solution is to use groups (http://symfony.com/blog/new-in-symfony-2-7-serialization-groups) like the following:
- add all properties of your user class to a group
- add properties you need from the group class to the same group except the `users` property to avoid the circular reference

**Comment by soyuka** [comment]
>If I understand well there is a circular reference in your exemple because of something like a users property in the UserGroup class: User -> UserGroup -> User -> UserGroup...

That's exactly it, because a UserGroup can have many users and a User can have many groups. The `UserGroup` entity is [the one from FOSUser](https://github.com/FriendsOfSymfony/FOSUserBundle/blob/master/Model/Group.php).

Thanks for the link, I knew I was missing something! I'll give it a try tomorrow and get back here to close/feedback.

**Comment by soyuka** [comment]
So, I've setup groups on my User entity ([gist](https://gist.github.com/soyuka/10b22a1422e9b4c29c55)). Note, I've put my properties in a `public` visibility just in case this was an issue, they should be `protected`.

First, I had to tweak the ViewHandler ([FOSRest#initResponse](https://github.com/FriendsOfSymfony/FOSRestBundle/blob/master/View/ViewHandler.php#L452)) to add groups in the serialization context.

Now, [AbstractNormalizer#getAllowedAttributes](https://github.com/symfony/Serializer/blob/master/Normalizer/AbstractNormalizer.php#L200) (wich is extended by the ObjectNormalizer) is not returning false, but I have no allowed attributes. `$this->classMetadataFactory->getMetadataFor($classOrObject)->getAttributesMetadata()` gets back an empty array. Should it not return attributes matching the context group (here `public`)?

Anyway, I'm not getting any circular reference errors but something like this:

```
[[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[],[]]
```

What am I doing wrong? 

Thanks!

**Comment by dunglas** [comment]
Can you create a Git repo with the working application? I'll give it a try.

Some inputs:
- Annotations works on `private` / `protected` annotations too (`ObjectNormalizer` use the PropertyAccess component.
- Never tried to use the new Serializer with FOSRest but it should work

**Comment by dunglas** [comment]
https://github.com/FriendsOfSymfony/FOSRestBundle/blob/master/View/ViewHandler.php#L452 is probably the problem.

To use groups, you must pass the groups you want to use in the third parameter of that method (ex: `$serializer->serialize($data, $format, array('groups' => array('private')));`).

**Comment by dunglas** [comment]
You should probably open a PR on FOSRest to add the support of the `serializerGroups` option to its `@View` annotation for the Symfony Serializer (it seems to only support JMSSerializer for now). ping @lsmith77

**Comment by soyuka** [comment]
Yes I know that's what I did already:

```
use Symfony\Component\Serializer\SerializerInterface as SymfonySerializerInterface;

            } else if($serializer instanceof SymfonySerializerInterface) {
                $context = $this->getSerializationContext($view);
                $groups = $context->attributes->get('groups')->getIterator();

                foreach($groups as $group) {
                    $this->exclusionStrategyGroups[] = $group[0];
                }

                $content = $serializer->serialize($data, $format, ['groups' => $this->exclusionStrategyGroups]);
            } else {
```

Just wrote before that my groups are now in the context as your serializer expects it:

```
['groups' => ['public']]
```

I'm currently trying the same with a non-circular entity.

**Comment by soyuka** [comment]
Suggestion: What about an annotation to help with `ignoredAttributes` too?

**Comment by dunglas** [comment]
I'm not a fan of having many way to do the same thing. You can ignore attributes using groups.

**Comment by soyuka** [comment]
There you go:

https://github.com/soyuka/testSymfonySerializer

To set it up:
```
composer update
app/console doctrine:schema:update --force
app/console doctrine:fixture:load
app/consoler server:run
```

**Comment by soyuka** [comment]
I've found my issue, I hadn't enabled annotations in the serializer configuration!

This should be added to the docs :).

`config.yml`:
```
framework:
    serializer: { enable_annotations: true }
```

I'm doing a pull request for the FOSRest bundle regarding the symfony serializer. Thanks @dunglas for your time.

**Comment by soyuka** [comment]
Now annotations are working and I can get rid of properties by selecting a different group. But, if I'm setting the same group on properties that had a circular reference, they are now returning an empty array.

See [this commit](https://github.com/soyuka/testSymfonySerializer/commit/d4685697c896b8144c566db54387367ed8d8e7c4) on the reproduction repository. 

`http://localhost:8000/api/v1/users` is now returning:
```
[{"id":1,"groups":[]}]
```

Should this not return the same error I had without groups annotations?

**Comment by dunglas** [comment]
Found the problem and fixed it in #14540.

Another thing to get your project working: don't forget to add the `@Groups` annotations on properties of the child entity  (`UserGroup` in your code).

**Comment by soyuka** [comment]
This indeed fixes this  :+1:, thanks!
